taichi/transforms/constant_fold.cpp

```diff
@@ -42,30 +47,86 @@ class ConstantFold : public BasicStmtVisitor {
         throw IRModified();
       }
     }
+#endif
+  }
+
+  struct BinaryEvaluatorId
+  {
+    BinaryOpType op;
+    DataType ret, lhs, rhs;
+
+    explicit operator int() const // make STL map happy
+    {
+      return (int)op | (int) ret << 8 | (int) lhs << 16 | (int) rhs << 24;
+    }
```

Comment:
You may use something like this (this is for `std::pair`; in our case, you may change the hash function in the following code to something you like, and `template` is not necessary):
```C++
struct hash_pair { 
    template <class T1, class T2> 
    size_t operator()(const pair<T1, T2>& p) const
    { 
        auto hash1 = hash<T1>{}(p.first); 
        auto hash2 = hash<T2>{}(p.second); 
        return hash1 ^ hash2; 
    } 
}; 
unordered_map<pair<int, int>, bool, hash_pair> um; 
```